src/main/java/com/braintreepayments/apollo_tracing_uploader/TraceProducer.java

```diff
@@ -1,18 +1,10 @@
 package com.braintreepayments.apollo_tracing_uploader;
 
-import java.util.function.Consumer;
-
 import mdg.engine.proto.Reports;
 
-public abstract class TraceProducer {
-  protected final FullTracesReportBuilder reportBuilder = new FullTracesReportBuilder();
-  protected final Consumer<Reports.ReportHeader.Builder> customizeReportHeader;
-  protected final Uploader uploader;
-
-  public TraceProducer(Consumer<Reports.ReportHeader.Builder> customizeReportHeader, Uploader uploader) {
-    this.customizeReportHeader = customizeReportHeader;
-    this.uploader = uploader;
-  }
+public interface TraceProducer {
```

Comment:
I do agree it feels a little awkward for this to be an abstract class, but the reason I did it was to enforce the need for a report header customizer consumer and an uploader in the constructor, because any "implementation" of `TraceProducer` is responsible for calling the consumer on the report header and uploading the report.

Would it make sense to add the interface as you've done, but then provide an `AbstractTraceProducer` implementation that contains this constructor? Then `ScheduledBatchingTraceProducer` could `extend AbstractTraceProducer`?